services/st_util/StUartBootloaderCommandHandler.hpp

```diff
@@ -0,0 +1,187 @@
+
+            void DataReceived() override;
+        };
+
+        class ReceiveBufferAction
+            : public ReceiveAction
+        {
+        public:
+            ReceiveBufferAction(StUartBootloaderCommandHandler& handler, infra::ByteRange& data);
+            ReceiveBufferAction(StUartBootloaderCommandHandler& handler, infra::ByteRange& data, const std::size_t nBytesTotal);
+
+            void DataReceived() override;
+
+        protected:
+            virtual void ExtractNumberOfBytes(infra::ByteInputStream& stream);
+
+        protected:
+            infra::ByteRange& data;
+            infra::Optional<std::size_t> nBytesTotal;
+            std::size_t nBytesReceived = 0;
+        };
+
+        class ReceiveSpecialBufferAction
+            : public ReceiveBufferAction
+        {
+        public:
+            using ReceiveBufferAction::ReceiveBufferAction;
+
+        protected:
+            void ExtractNumberOfBytes(infra::ByteInputStream& stream) override;
+        };
+
+        class TransmitAction
+            : public Action
+        {
+        public:
+            explicit TransmitAction(StUartBootloaderCommandHandler& handler);
+        };
+
+        class TransmitRawAction
+            : public TransmitAction
+        {
+        public:
+            TransmitRawAction(StUartBootloaderCommandHandler& handler, infra::ConstByteRange data);
+
+            void SendData() override;
+
+        private:
+            infra::ConstByteRange data;
+        };
+
+        class TransmitWithTwosComplementChecksum
+            : public TransmitAction
+        {
+        public:
+            TransmitWithTwosComplementChecksum(StUartBootloaderCommandHandler& handler, uint8_t data);
+
+            void SendData() override;
+
+        private:
+            uint8_t data;
+        };
+
+        class TransmitWithChecksumAction
+            : public TransmitAction
+        {
+        public:
+            TransmitWithChecksumAction(StUartBootloaderCommandHandler& handler, infra::ConstByteRange data);
+
+            void SendData() override;
+
+        private:
+            infra::ConstByteRange data;
+        };
+
+    private:
+        hal::SerialCommunication& serial;
+        infra::AutoResetFunction<void()> onInitialized;
+        infra::AutoResetFunction<void(infra::BoundedConstString reason)> onError;
+        infra::QueueForOneReaderOneIrqWriter<uint8_t>::WithStorage<257> queue;
+
+        infra::BoundedDeque<infra::PolymorphicVariant<Action, ReceiveAckAction, ReceiveBufferAction, ReceiveSpecialBufferAction, TransmitRawAction, TransmitWithTwosComplementChecksum, TransmitWithChecksumAction>>::WithMaxSize<12> commandActions;
+        infra::AutoResetFunction<void(), sizeof(StUartBootloaderCommandHandler*) + sizeof(infra::Function<void()>) + sizeof(infra::ByteRange)> onCommandExecuted;
+        infra::TimerSingleShot timeout;
+
+        std::array<uint8_t, 257> internalBuffer;
+        infra::ByteRange internalRange;
```

Comment:
These two variables would probably best be combined into one infra::BoundedVector